### PR TITLE
Feature 3.4/hotbackup devel merged

### PR DESCRIPTION
**Pull request by kvahed**

### CODE REVIEW

File: arangod/Cluster/ClusterMethods.cpp [hunk 1]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  VPackBuilder plan;
+
+  auto result = findLocalBackup(backupId, dbServers, plan);
+  if (!result.ok()) {
+    LOG_TOPIC(ERR, Logger::HOTBACKUP)
+      << "failed to find backup " << backupId << " on all db servers: "
+      << result.errorMessage();
+      return result;    
+  }
+  if (plan.slice() == VPackSlice::noneSlice()) {
```
Comment (neunhoef) [inline arangod/Cluster/ClusterMethods.cpp]: Use `isNone()` for better performance.

File: arangod/Cluster/ClusterMethods.cpp [hunk 2]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  std::this_thread::sleep_for(std::chrono::seconds(5));
+
+  // Restore all db servers
+  result = restoreOnDBServers(backupId, dbServers);
+  if (!result.ok()) {  // This is disaster!
+    return result;
+  }
+
+  // Proceed maintenance
+  result = controlMaintenanceFeature("proceed", backupId, dbServers);
```
Comment (neunhoef) [inline arangod/Cluster/ClusterMethods.cpp]: Not necessary, since the servers all restart.

File: arangod/Cluster/ClusterMethods.cpp [hunk 3]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  while (cc != nullptr && end < steady_clock::now()) {
+
+    CoordTransactionID coordTransactionID = TRI_NewTickServer();
+    std::string const lockUrl =
+      std::string("/_admin/hotbackup/lock?timeout=") + std::to_string(lockWait);
+
+    auto lockWaitEnd = steady_clock::now() + seconds((uint64_t)lockWait);
+
+    std::unordered_map<std::string, OperationID> lockRes;
+    for (auto const& dbServer : dbServers) {
```
Comment (neunhoef) [inline arangod/Cluster/ClusterMethods.cpp]: Could use `performRequests`?

File: arangod/Cluster/ClusterMethods.cpp [hunk 4]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+        }
+      } else {
+        std::stringstream s;
+        s << "Failed to send lock request ";
+        s << backupId + " to " + waitRes.serverID; 
+        LOG_TOPIC(ERR, Logger::HOTBACKUP) << s.str();
+        return arangodb::Result(TRI_ERROR_INTERNAL, s.str());
+      }
+    }
+
```
Comment (neunhoef) [inline arangod/Cluster/ClusterMethods.cpp]: Do we not have to cancel the locks on all servers, where we have succeeded, if any of the locks were not achieved until the timeout happens? Is using the timeout the right method here? Or should we rather have the other side return within a prescribed time, regardless on whether the lock has succeeded or not? This logic seems to be too simple.

File: arangod/Cluster/ClusterMethods.cpp [hunk 5]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  std::vector<ClusterCommRequest> requests;
+  
+  for (auto const& dbServer : dbServers) {
+    requests.emplace_back("server:" + dbServer, RequestType::POST, url, body);
+  }
+
+  // Perform the requests
+  size_t nrDone = 0;
+  cc->performRequests(
+    requests, CL_DEFAULT_TIMEOUT, nrDone, Logger::COMMUNICATION, false);
```
Comment (neunhoef) [inline arangod/Cluster/ClusterMethods.cpp]: Use `Logger::HOTBACKUP`.

File: arangod/Cluster/ClusterInfo.cpp [hunk 6]
```diff
@@ -3636,120 +3636,199 @@ arangodb::Result ClusterInfo::agencyReplan(VPackSlice const plan) {
...
+        builder.add("old", VPackValue("on")); } 
+    }
+  }
+
+  // Try to establish hot backup lock in agency. Result will always be 412.
+  // Question is: How 412?
+  auto result =
+    _agency.sendWithFailover(
+      arangodb::rest::RequestType::POST, (double)timeout, writeURL, builder.slice());
+  if (!result.successful() &&
```
Comment (neunhoef) [inline arangod/Cluster/ClusterInfo.cpp]: It seems this && must be an ||

File: arangod/Cluster/ClusterInfo.cpp [hunk 7]
```diff
@@ -3636,120 +3636,199 @@ arangodb::Result ClusterInfo::agencyReplan(VPackSlice const plan) {
...
+  auto rv = VPackParser::fromJson(result.bodyRef());
+
+  if (!rv->slice().isObject() || !rv->slice().hasKey("results") ||
+      !rv->slice().get("results").isArray()) {
+    return arangodb::Result(
+      TRI_ERROR_HOT_BACKUP_INTERNAL, "invalid agency result while acuiring backup lock" );    
+  }
+  auto ar = rv->slice().get("results");
+
+  uint64_t first = ar[0].getNumber<uint64_t>();
```
Comment (neunhoef) [inline arangod/Cluster/ClusterInfo.cpp]: What if the array length is not >= 2?

File: arangod/Cluster/ClusterInfo.cpp [hunk 8]
```diff
@@ -3636,120 +3636,199 @@ arangodb::Result ClusterInfo::agencyReplan(VPackSlice const plan) {
...
+  uint64_t second = ar[1].getNumber<uint64_t>();
+  
+  if (first == 0 && second == 0) { // tough luck
+    return arangodb::Result(
+      TRI_ERROR_HOT_BACKUP_INTERNAL,
+      "preconditions failed while trying to acquire backup lock in the agency");
+  } else if (first > 0) {          // Supervision was on
```
Comment (neunhoef) [inline arangod/Cluster/ClusterInfo.cpp]: This code does not make sense.

File: arangod/Cluster/ClusterInfo.cpp [hunk 9]
```diff
@@ -3594,12 +3599,159 @@ arangodb::Result ClusterInfo::getShardServers(ShardID const& shardId,
 
 arangodb::Result ClusterInfo::agencyDump(std::shared_ptr<VPackBuilder> body) {
 
-  AgencyCommResult dump = _agency.dump();
+  AgencyCommResult dump = _agency.getValues("arango/Plan");
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: Of course it does. Fixed.

File: arangod/Cluster/ClusterInfo.cpp [hunk 10]
```diff
@@ -3594,12 +3599,159 @@ arangodb::Result ClusterInfo::getShardServers(ShardID const& shardId,
...
+  AgencyWriteTransaction planTransaction(
+    std::vector<AgencyOperation>
+    {AgencyOperation("arango/Plan/Collections", AgencyValueOperationType::SET,
+                     plan.get("Collections")),
+     AgencyOperation("arango/Plan/Databases", AgencyValueOperationType::SET,
+                     plan.get("Databases"))});
+
+  AgencyCommResult r = _agency.sendTransactionWithFailover(planTransaction);
+  if (!r.successful()) {
+    arangodb::Result result (
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: The return type is not `arangodb::Result`, but `arangodb::AgencyCommResult`

File: arangod/Cluster/ClusterInfo.cpp [hunk 11]
```diff
@@ -3594,12 +3599,159 @@ arangodb::Result ClusterInfo::getShardServers(ShardID const& shardId,
...
+  }
+  
+  return arangodb::Result();
+  
+}
+
+
+
+arangodb::Result ClusterInfo::agencyHotBackupUnlock(std::string const& backupId) {
+  return Result();
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: No longer. Was victim of complete rewrite on Friday.

File: arangod/Cluster/ClusterInfo.cpp [hunk 12]
```diff
@@ -3594,12 +3599,159 @@ arangodb::Result ClusterInfo::getShardServers(ShardID const& shardId,
...
+  //    Else report error
+  // 2. Wait For /arango/Target/HotBackup/UUID = 1
+  // 3. For 30 seconds. Send lock commands to all db servers timeout 2.
+  //    If lock obtained -> Send create
+  //    Else report error
+
+  std::string const backupURL("Target/HotBackup/" + backupId);
+
+  AgencyComm ac;
+  VPackBuilder lock;
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: The goal is to write a transaction in for `[[{},{}],[{},{}]`.

File: arangod/Cluster/ClusterMethods.cpp [hunk 13]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  std::vector<ClusterCommRequest> requests;
+  for (auto const& dbServer : dbServers) {
+    requests.emplace_back(
+      "server:" + dbServer, RequestType::POST, url, body);
+  }
+
+  // Perform the requests
+  size_t done = 0;
+  cc->performRequests(
+    requests, CL_DEFAULT_TIMEOUT, done, Logger::COMMUNICATION, false);
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: Done

File: arangod/Cluster/ClusterMethods.cpp [hunk 14]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+
+    auto res = req.result;
+    int commError = handleGeneralCommErrors(&res);
+    if (commError != TRI_ERROR_NO_ERROR) {
+      return arangodb::Result(
+        commError,
+        std::string("Communication error while getting listt of backups from ")
+        + req.destination);
+    }
+    TRI_ASSERT(res.answer != nullptr);
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: The rest of the functions in this file depend on this?

File: arangod/Cluster/ClusterMethods.cpp (hunk 1 above)
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: Done

File: arangod/Cluster/ClusterMethods.cpp (hunk 2 above)
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: True. Done

File: arangod/Cluster/ClusterMethods.cpp (hunk 3 above)
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: Done

File: arangod/Cluster/ClusterMethods.cpp (hunk 4 above)
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: Done

File: arangod/Cluster/ClusterMethods.cpp (hunk 5 above)
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: Done

File: arangod/Cluster/ClusterInfo.cpp (hunk 6 above)
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: Correct. Commented less obscruely

File: arangod/Cluster/ClusterInfo.cpp (hunk 7 above)
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: addressed

File: arangod/Cluster/ClusterInfo.cpp (hunk 8 above)
Comment (kvahed) [reply arangod/Cluster/ClusterInfo.cpp]: OMG!

File: arangod/Cluster/MaintenanceWorker.cpp [hunk 15]
```diff
@@ -56,7 +56,7 @@ MaintenanceWorker::MaintenanceWorker(arangodb::MaintenanceFeature& feature,
 void MaintenanceWorker::run() {
   bool more(false);
 
-  while (eSTOP != _loopState && !_feature.isShuttingDown()) {
+  while (eSTOP != _loopState && !_feature.isShuttingDown() && !_feature.isPaused()) {
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/MaintenanceWorker.cpp]: Done

File: arangod/Cluster/ClusterMethods.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: Done throughout the code above

File: arangod/Cluster/ClusterMethods.cpp [hunk 16]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+    auto resBody = res.answer->toVelocyPackBuilderPtrNoUniquenessChecks();
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: just kept to above code

File: arangod/Cluster/ClusterMethods.cpp [hunk 17]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+      + std::to_string(planServers.length()) +
+      ") and in this cluster (" + std::to_string(dbServers.size()) + ") do not match");
+  }
+
+  // Clear match container
+  match.clear();
+
+  // Local copy of my servers
+  std::unordered_set<std::string> localCopy;
+  std::copy(dbServers.begin(), dbServers.end(),
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: No, I would expect equivalent.

File: arangod/Cluster/ClusterMethods.cpp [hunk 18]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+            std::inserter(localCopy, localCopy.end()));
+
+  // Skip all direct matching names in pair and remove them from localCopy
+  std::unordered_set<std::string>::iterator it; 
+  for (auto const& planned : VPackArrayIterator(planServers)) {
+    auto const plannedStr = planned.copyString();
+    if ((it = std::find(localCopy.begin(), localCopy.end(), plannedStr)) != localCopy.end()) {
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: hippster c++ it is

File: arangod/Cluster/ClusterMethods.cpp [hunk 19]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  std::string const url = "_admin/actions";
+  auto body = std::make_shared<std::string>(builder.toJson());
+  for (auto const& dbServer : dbServers) {
+    requests.emplace_back("server:" + dbServer, RequestType::POST, url, body);
+  }
+  
+  // Perform the requests
+  size_t done = 0;
+  cc->performRequests(
+    requests, CL_DEFAULT_TIMEOUT, done, Logger::COMMUNICATION, false);
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: done

File: arangod/Cluster/ClusterMethods.cpp [hunk 20]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+  return arangodb::Result();
+
+}
+
+
+arangodb::Result applyDBServerMatchesToPlan(
+  VPackSlice const plan, std::unordered_map<ServerID,ServerID> const& matches,
+  VPackBuilder& newPlan) {
+
+  std::function<void(
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: cannot be declared `auto`, cause a deduced type cannot appear in own initialiser

File: arangod/Cluster/ClusterMethods.cpp [hunk 21]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+      }
+    } else if (s.isArray()) {
+      VPackArrayBuilder a(&newPlan);
+      for (auto const& it : VPackArrayIterator(s)) {
+        replaceDBServer(it, matches);
+      }
+    } else {
+      if (s.isString()) {
+        bool replaced = false;
+        for (auto const& match : matches) {
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: changed to map

File: arangod/Cluster/ClusterMethods.cpp [hunk 22]
```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
...
+    }
+
+    if (!resSlice.hasKey("server") || !resSlice.hasKey("hotbackups")
+        || !resSlice.get("hotbackups").isArray()) {
+      return arangodb::Result(
+        TRI_ERROR_HOT_BACKUP_INTERNAL,
+        std::string("invalid response from ") + req.destination);
+    }
+
+    for (auto const id : VPackArrayIterator(resSlice.get("hotbackups"))) {
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Cluster/ClusterMethods.cpp]: done

**Pull request merged**